### DISCUSSION TITLE
Normalize color band

### DISCUSSION THREAD

**Discussion by mtiessen1175** [libvips/php-vips#250] [Q&A] [answered]
Hi everyone,

I'm trying to do a simple normalization of a gray-scale image color-band from negative range to a desired range of 0 to 255 (see code below).
<details>
<summary> code setup </summary>

```php
     /**
     * Generate tiles for the object and put them to temporary storage.
     *
     * @param File $file
     * @param string $path Path to the cached image file.
     */
    public function generateTiles($file, $path)
    {
        $vipsImage = VipsImage::newFromFile($path);
        $min = $vipsImage->min();
        $max = $vipsImage->max();

        if($min < 0 || $min > 255 || $max < 0 || $max > 255) {
            $vipsImage = $vipsImage->subtract($min)->divide($max - $min)->multiply(255);
            $vipsImage->dzsave($this->tempPath, [
                'layout' => 'zoomify',
                'container' => 'fs',
                'strip' => true,
            ]);
        }
    }
```
</details>

The example images I use are geoTIFF files which have a negative color-band range (e.g. this [example .tif](https://github.com/user-attachments/files/17620287/example_tif.zip) with `min: -4123.31` and `max: -4061.37`; information derived from QGIS, see screenshot below). 
I am able to get the max-value of the image correctly via `$vipsImage->max()`. The `$vipsImage->min()` returns -99999. However, I would like to get the min-value of the true image pixels (-4123.31), as with the current setup, I get a black&white image with binary values (either 0 or 255) instead of a gray-scale image. 
![QGIS_information_tab](https://github.com/user-attachments/assets/729544c9-0f4a-4bbc-b086-c96d1ed29587)

So my question is, how could I get the min-value of the true pixel values of the geoTIFF and disregard the NoData pixel values.

Any help to achieve this is greatly appreciated! Thanks in advance for your time and effort.

**Comment by jcupitt** [comment, marked as answer]
Hello @mtiessen1175,

Do you have a sample image you could share?

If you want to find the min and exclude -99999, it's probably easiest to set the -99999s to 0 (or some very high value?) before searching for min. Perhaps (untested):

```php
$min = $image->equal(-99999)->ifthenelse(0, $image)->min();
```

You can speed it up with `stats()`. This will make a single pass over the image and find max and min together.

You could also call `linear()` directly. This computes `y = a * x + b`, so you can replace your -, /, * with a single call.

**Reply by mtiessen1175** [reply]
@jcupitt, thanks for demonstrating how to "exclude" data values and also for the suggested code improvements. The easy fix did the trick, i.e. the color normalization works as expected now that the min and max are set correctly.

Thank you very much for your help!
